src/IdentityServer/Validation/Default/TokenRequestValidator.cs

```diff
@@ -335,7 +335,7 @@ private async Task<TokenRequestValidationResult> ValidateAuthorizationCodeReques
 
         _validatedRequest.AuthorizationCodeHandle = code;
 
-        var authZcode = await _authorizationCodeStore.GetAuthorizationCodeAsync(code);
+        var authZcode = await _authorizationCodeStore.GetAndDeleteAuthorizationCodeAsync(code);
```

Comment:
This _may_ not be desired behaviour. This will result in the deletion of the persisted grant in the following scenarios:
```csharp
authZcode.ClientId != _validatedRequest.Client.ClientId or
(authZcode.DPoPKeyThumbprint.IsPresent() && _validatedRequest.ProofType != ProofType.DPoP) or
(authZcode.DPoPKeyThumbprint.IsPresent() && _validatedRequest.ProofKeyThumbprint != authZcode.DPoPKeyThumbprint)
```
This would be a breaking change for those that rely on the grant to not be deleted in these scenarios.
I would argue that in the case of DPoP failing (and therefore likely a code having been leaked), it would be useful to have that grant revoked.